### ISSUE TITLE
Can not set value to IntVar

### ISSUE THREAD

**Issue by kunalsapru** [chocoteam/choco-solver#551]
Hi Team,

I have used ChocoSolver 2 in past and now migrating to ChocoSolver 4.
I want to check whether a particular solution is valid or not?
Earlier there were methods to setValue of IntVar and then solver.checkSolution would check feasibility of the solution. This would give me the exact constraint that failed, but now I can't find any method to set value for variables and check for a particular solution.
An early reply will be highly appreciated.

Thanks,
Kunal

**Comment by jgFages** [comment]
Hi,

var.instantiateTo(42,aCause.NULL);
model.getSolver().propagate(); // with try catch
-> throughs exception in case of inconsistencies

**Comment by kunalsapru** [comment]
Thank you for your response. Using instantiateTo() was giving an exception that the variable has already been instantiated, but I got the idea. Now, I'm doing separate checks for variable's domain using var.contains(value) and for constraint checking, I'm declaring a new variable using just IntVar v1 = model.intVar(value) and then posting the constraints.

Thanks for quick response.

**Comment by kunalsapru** [comment]
Hi,

I have another question. How can we add a new constraint and attach it to the model later ? 
For eg, how can we declare an independent constraint like Constraint c = new Member(v, from, to) and attach it to model? 
In the above comment, although I'm able to create a new variable but I'm not able to apply the same constraint on it that was applied before. The best solution for me would be to modify the value of an existing IntVar.

**Comment by cprudhom** [comment]
You can `post(c)` it whenever you want, it then becomes permanent.
Otherwise, you can `postTemp(c)` it, it will be unpost on backtrack.

API are reachable from `Model`.

**Comment by kunalsapru** [comment]
Hi,
Thanks for your reply.
Let me rephrase my question in a more precise and simple way:
Step1 ) I have created a model.
Step2 ) I have created variables like IntVar with their domain values.
Step3 ) I have associated constraints against all these variables.
Step4 ) I have a solution from user and I want to check whether it satisfies all the constraints that I have defined. How would I do it?

Any help would be highly appreciated. 
Earlier, in ChocoSolver2, we were just setting the value of the varibles using setVal() and then using solver.checkSolution()  method. Is there a same way to do this?

Many Thanks,
Kunal

**Comment by cprudhom** [comment]
What you have to do is something like that:
```java
Model model = new Model();
// declare variables and constraints here
Solver solver = molde.getSolver();
while(/*stop criterion*/){
  model.getEnvironment().worldPush(); 
  try{
    // instantiate a variable here
    solver.propagate();
  }catch (ContradictionException sex){
    solver.getEngine().flush();
  }
  model.getEnvironment().worldPop();
}
```

**Comment by kunalsapru** [comment]
Hi,
Thanks for your answer. When I try to instantiate a variable, I get variable already instantiated exception.
Let me give you a code snippet:
`        Model cs4model = new Model();
        IntVar v = cs4model.intVar(name, lowerBound, upperBound);
//name, lowerBound, upperBound are fetched form user program
        addConstraint(cs4model.arithm(v, ">=", 50));
//Here I want to check for a particular solution by iterating over a map variable "solutionMap" which //contains variable name and value
        Solver solver = cs4model.getSolver();
        boolean solveStatus = true;
        cs4model.getEnvironment().worldPush();
        int val;
        IntVar[] intVars = cs4model.retrieveIntVars(true);
        for(IntVar var : intVars){
            if(solutionMap.containsKey(var.getName())){
                val = solutionMap.get(var.getName());
                try {
                    IntVar v1 = cs4model.intVar(var.getName(), val);
//or var = instantiateTo(val, null);
// I get exception in both cases
                    solver.propagate();
                } catch (ContradictionException e) {
                    solveStatus = false;
                    e.printStackTrace();
                }
            }
        }
        cs4model.getEnvironment().worldPop();

`

**Comment by cprudhom** [comment]
The worldPush/pop should be *inside* the loop, not outside.
Double check my pseudo code ;)

**Comment by kunalsapru** [comment]
Many thanks. This works like a charm. :+1:

**Comment by kunalsapru** [comment]
```java
        if(domainCheck) {
            //Check for constraints, only in case of no domain error(s)
            for(IntVar var : intVars){
                if(solution.containsKey(var.getName())){
                    val = solution.get(var.getName());
                    for(Constraint c : constraints) {
                        cs4model.getEnvironment().worldPush();
                        cs4model.post(c);
                            try {
                                var.instantiateTo(val, null);
                                solver.propagate();
                            } catch (ContradictionException e) {
                                solver.getEngine().flush();
                                domainCheck = false;
                                result.add("Constraint failed for "+var.getName()+" with value "+val+" --> "+c.toString());
                            }
                        cs4model.unpost(c);
                        cs4model.getEnvironment().worldPop();
                        if(!(domainCheck)){
                            break;
                        }
                    }
                }
            }
        }
```
